### ISSUE TITLE
dnsdist expungeByName() is slow

### ISSUE THREAD

**Issue by EarthDevour** [PowerDNS/pdns#12438]
- Program:dnsdist 
 - Issue type: Feature request

### Short description

PacketCache:expungeByName  is slow
PacketCache:expungeByName  takes only one newDNSName()

### Usecase

We are using `expungeByName` to purge records that have been recently changed by customers, however in busy environemnts (5 mil packet cache) dnsdist can't keep up with requests to purge. At the moment we are in the deadloop of ever increasing events of "to be purged" on rabbitmq(it is used to deliver messages what needs to be purged)

### Description

We have talked a bit in chat with habbie about this, for that time ideas was 

- having index in packet cache for faster "queries"
- having PacketCache:expungeByName take multiple dnsnames

any other suggestion?

**Comment by rgacogne** [comment]
Right, 5 millions entries is a huge packet cache, and at the moment removing by name means walking through all entries present in the cache, so it _will_ hurt. The design is not optimized for that.

Adding an index would require a completely different design of the packet cache, likely based on a `multi_index_container` like what we do in the auth and recursor. It would also significantly improve the cleaning performance, which is not great at the moment. Of course it would means significantly slower cache lookups, so I don't think such an implementation should replace the existing one, but the current design makes it easy to have two different cache implementations and let the user pick the one they want.

Having `PacketCache:expungeByName` would not be hard, and we already have a pull request for that: https://github.com/PowerDNS/pdns/issues/7157 Unfortunately no-one seemed to have cared enough to contribute the code, so far.
